Should study [open-context-jupyter/Open Context Measurements.ipynb at master · ekansa/open-context-jupyter](https://github.com/ekansa/open-context-jupyter/blob/master/notebooks/Open%20Context%20Measurements.ipynb)

May make use of [digitalbazaar/pyld: JSON-LD processor written in Python](https://github.com/digitalbazaar/pyld)

In [5]:
from itertools import islice
import json
from urllib.parse import urljoin
from urllib.parse import urlencode, quote_plus

import requests
from requests.packages.urllib3.util.retry import Retry
from requests.adapters import HTTPAdapter

from pyld import jsonld

import pandas as pd
import numpy as np
from pandas import DataFrame, Series, Index

In [17]:
def filter_none(d):
    return dict([(k,v) for (k,v) in d.items() if v is not None])


class OpenContextAPI(object):
    def __init__(self, api_url="https://opencontext.org/"):
        self.api_url = api_url
        self.session = requests.Session()
        retries = Retry(total=5,
                backoff_factor=0.1,
                status_forcelist=[ 500, 502, 503, 504 ])
        self.session.mount(self.api_url, HTTPAdapter(max_retries=retries))
        
    def projects_search(self, start=None, stop=None, rows=None):

        kw = filter_none({
            'start': start,
            'stop': stop,
            'rows': rows
        })
        
        url = urljoin(self.api_url, "projects-search.json") + "?" + urlencode(kw, quote_via=quote_plus)
    
        while url:
            r = self.session.get(url)
            if r.status_code == 200:
                features = r.json().get('features', [])
                yield from features
            else:
                # to do: figure what to do when non-200 status code
                pass
            url = r.json().get('next-json', None)

                


In [19]:
oc = OpenContextAPI()
features = oc.projects_search()

projects_df = DataFrame(list(islice(features,None)))
projects_df

category                                           geometry  \
0   oc-api:geo-record  {'id': '#geo-rec-geom-1-of-109', 'type': 'Poin...   
1   oc-api:geo-record  {'id': '#geo-rec-geom-2-of-109', 'type': 'Poin...   
2   oc-api:geo-record  {'id': '#geo-rec-geom-3-of-109', 'type': 'Poin...   
3   oc-api:geo-record  {'id': '#geo-rec-geom-4-of-109', 'type': 'Poin...   
4   oc-api:geo-record  {'id': '#geo-rec-geom-5-of-109', 'type': 'Poin...   
5   oc-api:geo-record  {'id': '#geo-rec-geom-6-of-109', 'type': 'Poin...   
6   oc-api:geo-record  {'id': '#geo-rec-geom-7-of-109', 'type': 'Poin...   
7   oc-api:geo-record  {'id': '#geo-rec-geom-8-of-109', 'type': 'Poin...   
8   oc-api:geo-record  {'id': '#geo-rec-geom-10-of-109', 'type': 'Poi...   
9   oc-api:geo-record  {'id': '#geo-rec-geom-11-of-109', 'type': 'Poi...   
10  oc-api:geo-record  {'id': '#geo-rec-geom-12-of-109', 'type': 'Poi...   
11  oc-api:geo-record  {'id': '#geo-rec-geom-13-of-109', 'type': 'Poi...   
12  oc-api:geo-record  {'id': '#geo-rec-geom-15-of-109', 'type': 'Poi...   
13  oc-api:geo-record  {'id': '#geo-rec-geom-17-of-109', 'type': 'Poi...   
14  oc-api:geo-record  {'id': '#geo-rec-geom-18-of-109', 'type': 'Poi...   
15  oc-api:geo-record  {'id': '#geo-rec-geom-19-of-109', 'type': 'Poi...   
16  oc-api:geo-record  {'id': '#geo-rec-geom-21-of-109', 'type': 'Poi...   
17  oc-api:geo-record  {'id': '#geo-rec-geom-22-of-109', 'type': 'Poi...   
18  oc-api:geo-record  {'id': '#geo-rec-geom-24-of-109', 'type': 'Poi...   
19  oc-api:geo-record  {'id': '#geo-rec-geom-25-of-109', 'type': 'Poi...   
20  oc-api:geo-record  {'id': '#geo-rec-geom-26-of-109', 'type': 'Poi...   
21  oc-api:geo-record  {'id': '#geo-rec-geom-27-of-109', 'type': 'Poi...   
22  oc-api:geo-record  {'id': '#geo-rec-geom-29-of-109', 'type': 'Poi...   
23  oc-api:geo-record  {'id': '#geo-rec-geom-31-of-109', 'type': 'Poi...   
24  oc-api:geo-record  {'id': '#geo-rec-geom-32-of-109', 'type': 'Poi...   
25  oc-api:geo-record  {'id': '#geo-rec-geom-33-of-109', 'type': 'Poi...   
26  oc-api:geo-record  {'id': '#geo-rec-geom-34-of-109', 'type': 'Poi...   
27  oc-api:geo-record  {'id': '#geo-rec-geom-35-of-109', 'type': 'Poi...   
28  oc-api:geo-record  {'id': '#geo-rec-geom-36-of-109', 'type': 'Poi...   
29  oc-api:geo-record  {'id': '#geo-rec-geom-37-of-109', 'type': 'Poi...   
..                ...                                                ...   
48  oc-api:geo-record  {'id': '#geo-rec-geom-61-of-109', 'type': 'Poi...   
49  oc-api:geo-record  {'id': '#geo-rec-geom-62-of-109', 'type': 'Poi...   
50  oc-api:geo-record  {'id': '#geo-rec-geom-63-of-109', 'type': 'Poi...   
51  oc-api:geo-record  {'id': '#geo-rec-geom-64-of-109', 'type': 'Poi...   
52  oc-api:geo-record  {'id': '#geo-rec-geom-65-of-109', 'type': 'Poi...   
53  oc-api:geo-record  {'id': '#geo-rec-geom-66-of-109', 'type': 'Poi...   
54  oc-api:geo-record  {'id': '#geo-rec-geom-68-of-109', 'type': 'Poi...   
55  oc-api:geo-record  {'id': '#geo-rec-geom-69-of-109', 'type': 'Poi...   
56  oc-api:geo-record  {'id': '#geo-rec-geom-70-of-109', 'type': 'Poi...   
57  oc-api:geo-record  {'id': '#geo-rec-geom-71-of-109', 'type': 'Poi...   
58  oc-api:geo-record  {'id': '#geo-rec-geom-73-of-109', 'type': 'Poi...   
59  oc-api:geo-record  {'id': '#geo-rec-geom-75-of-109', 'type': 'Poi...   
60  oc-api:geo-record  {'id': '#geo-rec-geom-80-of-109', 'type': 'Poi...   
61  oc-api:geo-record  {'id': '#geo-rec-geom-83-of-109', 'type': 'Poi...   
62  oc-api:geo-record  {'id': '#geo-rec-geom-84-of-109', 'type': 'Poi...   
63  oc-api:geo-record  {'id': '#geo-rec-geom-85-of-109', 'type': 'Poi...   
64  oc-api:geo-record  {'id': '#geo-rec-geom-86-of-109', 'type': 'Poi...   
65  oc-api:geo-record  {'id': '#geo-rec-geom-88-of-109', 'type': 'Poi...   
66  oc-api:geo-record  {'id': '#geo-rec-geom-89-of-109', 'type': 'Poi...   
67  oc-api:geo-record  {'id': '#geo-rec-geom-90-of-109', 'type': 'Poi...   
68  oc-api:geo-record  {'id': '#geo-rec-geom-91-of-109', 'typ

I don't understand why we get only 78 results and not 109.